docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,548 @@
+  meshservice:mesh/mesh-1:name/labeled-service:
+    conf: merge($conf1, $conf2, $conf4)
+  mesh:name/mesh-1:
+    conf: $conf1
+```
+
+Note that the structure doesn't contain `other-service`, but instead it contains entry for `mesh-1`.
+The code that computes confs should handle the absence of `other-service` and replace it with conf of the `mesh-1`.
+
+### Inspect API
+
+When it comes to the Inspect API, we're going to add a new field to `InspectRule`
+
+```yaml
+InspectRule:
+  type: object
+  required: [type]
+  properties:
+    type:
+      type: string
+      example: MeshRetry
+      description: the type of the policy
+    proxyRule:
+      description: a rule that affects the entire proxy
+      $ref: '#/components/schemas/ProxyRule'
+    toRules: # should be removed eventually
+      type: array 
+      description: a set of rules for the outbounds of this proxy
+      items:
+        $ref: '#/components/schemas/Rule'
+    toResourceRules: # new field
+      type: array
+      items:
+        $ref: '#/components/schemas/ResourceRule'
+    fromRules:
+      type: array
+      description: a set of rules for each inbound of this proxy
+      items:
+        $ref: '#/components/schemas/FromRule'
+    warnings:
+      type: array
+      description: a set of warnings to show in policy matching
+      example: ["Mesh is not Mtls enabled this policy will have no effect"]
+      items:
+        type: string
+Rule:
+  type: object
+  required: [matchers, conf, origin]
+  properties:
+    matchers:
+      type: array
+      items:
+        $ref: '#/components/schemas/RuleMatcher'
+    conf:
+      description: The actual conf generated
+      type: object
+      x-go-type: 'interface{}'
+    origin:
+      type: array
+      items:
+        $ref: '#/components/schemas/Meta'
+ResourceRule: # new type
+  type: object
+  required: [resourceMeta, conf, origin]
+  properties:
+    resourceMeta:
+      $ref: '#/components/schemas/Meta'
+    resourceSectionName:
+      type: string
+    conf:
+      description: The actual conf generated
+      type: object
+      x-go-type: 'interface{}'
+    origin:
+      type: array
+      description: |
+        The list of policies that contributed to the 'conf'. The order is important as it reflects 
+        in what order confs were merged to get the resulting 'conf'. 
+      items:
+        type: object
+        properties:
+          resourceMeta:
+            $ref: '#/components/schemas/Meta'
+          toIdx:
+            description: index of the to-item in the policy
+            type: integer
+```
+
+The example of the Inspect API response for the given MeshTimeout policy would be:
+
+```yaml
+kind: MeshTimeout
+metadata:
+  name: mt-1
+  namespace: client-ns
+  labels:
+    kuma.io/mesh: mesh-1
+spec:
+  to:
+    - targetRef:
+        kind: Mesh
+      conf: conf1
+    - targetRef:
+        kind: MeshService
+        name: backend
+        namespace: finance
+      conf: conf2
+```
```

Comment:
Hmm one thing I've just realised it that the one above is equivalent to:

```yaml
kind: MeshTimeout
metadata:
  name: mt-1
  namespace: client-ns
  labels:
    kuma.io/mesh: mesh-1
spec:
  to:
    - targetRef:
        kind: Mesh
      conf: conf1
---
kind: MeshTimeout
metadata:
  name: mt-1
  namespace: client-ns
  labels:
    kuma.io/mesh: mesh-1
spec:
  to:
    - targetRef:
        kind: MeshService
        name: backend
        namespace: finance
      conf: conf2
```

Is this intentional? Did we play it out? I guess it makes sense but I want to make sure we've thourougly thought about this

Reply:
Yes, it's intentional, the reason why users would prefer a single policy with multiple to-items is to have configurations for different services in the same consumer policy.

Reply:
Yes, it's intentional, the reason why users would prefer a single policy with multiple to-items is to have configurations for different services in the same consumer policy.